homework/index.js

```diff
@@ -0,0 +1,206 @@
+class ShoppingMarketError extends Error {
+    static ERROR_CODE = {
+        EXPIRED_ALIMENT: 'EXPIRED_ALIMENT',
+        NOT_IN_CART: 'NOT_IN_CART',
+        INSUFFICIENT_FUNDS: 'INSUFFICIENT_FUNDS'
+    };
+
+    static ERROR_MESSAGE = {
+        [ShoppingMarketError.ERROR_CODE.EXPIRED_ALIMENT]:
+            'Aliment is expired',
+
+        [ShoppingMarketError.ERROR_CODE.NOT_IN_CART]:
+            'Product is not in the shopping cart',
+
+        [ShoppingMarketError.ERROR_CODE.INSUFFICIENT_FUNDS]:
+            'Customer has insufficient funds'
+    };
+
+    constructor({
+        code,
+        message = ShoppingMarketError.ERROR_MESSAGE[code],
+        data
+    }) {
+        super(message);
+
+        this.code = code;
+        this.data = data;
+    }
+
+    toString() {
+        return JSON.stringify({
+            code: this.code,
+            message: this.message,
+            data: this.data
+        }, null, 4) + '\n' + this.stack;
+    }
+}
+
+class Person {
+    #name;
+
+    constructor(name) {
+        this.#name = name;
+    }
+
+    get name() {
+        return this.#name;
+    }
+}
+
+class Customer extends Person {
+    #money;
+    #shoppingCart;
+
+    constructor({ name = 'unknown', money }) {
+        super(name);
+
+        this.#money = money;
+        this.#shoppingCart = [];
+    }
+
+    get money() {
+        return this.#money;
+    }
+
+    get shoppingCart() {
+        return this.#shoppingCart;
+    }
+
+    set money(money) {
+        this.#money = money;
+    }
+
+    addToCart(product) {
+        if (product instanceof Product === false) {
+            throw new TypeError('product must be a Product');
+        }
+
+        if (product instanceof Aliment && product.expirationDate < Date.now()) {
```

Comment:
Take care of the operator precedence. In this case, `&&` is evaluated before `instanceof` operator if we don't use brackets.

Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table

Reply:
You are right. Unfortunately, my brain mechanically parsed the table from MDN using the default order in such tables.

Namely, operators to be listed in descending precedence (e.g. https://en.cppreference.com/w/c/language/operator_precedence). 

So the numbering from MDN caused my confusion, although is debatable which numbering schema is more natural (e.g. I think of priority 0 as the most prioritary).